### PR TITLE
[management] add governance command

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

This commit adds add_validator/remove_validator command to the
operational tooling.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Adopt smoke tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: config/management/operational/src/command.rs [hunk 1]
```diff
@@ -48,6 +56,8 @@ impl std::fmt::Display for CommandName {
             CommandName::RotateFullNodeNetworkKey => "rotate-fullnode-network-key",
             CommandName::RotateValidatorNetworkKey => "rotate-validator-network-key",
             CommandName::ValidateTransaction => "validate-transaction",
+            CommandName::AddValidator => "add-validator",
+            CommandName::RemoveValidator => "remove-validator",
```
Comment (davidiw) [inline config/management/operational/src/command.rs:60]: gaah alphabetize

File: config/management/operational/src/governance.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+pub struct AddValidator {
+    #[structopt(long, help = "JSON-RPC Endpoint (e.g. http://localhost:8080")]
+    host: String,
+    #[structopt(long)]
+    chain_id: ChainId,
+    #[structopt(long, help = "The validator address to be added")]
+    address_to_add: AccountAddress,
```
Comment (davidiw) [inline config/management/operational/src/governance.rs:24]: maybe validator_address?

File: config/management/operational/src/command.rs [hunk 3]
```diff
@@ -103,6 +115,20 @@ impl Command {
...
+            _ => Err(self.unexpected_command(CommandName::AddValidator)),
+        }
+    }
+
+    pub fn remove_validator(self) -> Result<TransactionContext, Error> {
+        match self {
+            Command::RemoveValidator(cmd) => cmd.execute(),
+            _ => Err(self.unexpected_command(CommandName::RemoveValidator)),
+        }
+    }
```
Comment (gregnazario) [inline config/management/operational/src/command.rs:130]: @davidiw Any idea on how we could shrink this boilerplate stuff that we've put for every command?

File: config/management/operational/src/governance.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    #[structopt(flatten)]
+    libra_root: ValidatorBackend,
+}
+
+impl AddValidator {
+    pub fn execute(self) -> Result<TransactionContext, Error> {
+        let client = JsonRpcClientWrapper::new(self.host);
+        if let Err(e) = client.validator_config(self.address_to_add) {
+            return Err(Error::JsonRpcReadError("validator-config", e.to_string()));
+        }
```
Comment (gregnazario) [inline config/management/operational/src/governance.rs:34]: I'm pretty sure the error coming out of `validator_config` is the correct error at this point and you should be able to just put a `?` at the end.

File: config/management/operational/src/governance.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            constants::GAS_UNIT_PRICE,
+            constants::GAS_CURRENCY_CODE.to_owned(),
+            RealTimeService::new().now() + constants::TXN_EXPIRATION_SECS,
+            self.chain_id,
+        );
+        let mut storage =
+            StorageWrapper::new(self.libra_root.name(), &self.libra_root.validator_backend)?;
+        let signed_txn = storage.sign(LIBRA_ROOT_KEY, "remove-validator", txn)?;
+        client.submit_transaction(signed_txn)
+    }
```
Comment (gregnazario) [inline config/management/operational/src/governance.rs:86]: Nit: Not for you necessarily, but in my next refactor I want to refactor this out, as the transaction, signing, and submission are all the same, except the `script` and the `signing name`.  It's also basically the same as in validator config:

https://github.com/libra/libra/blob/master/config/management/src/validator_config.rs#L97-L111

File: config/management/src/json_rpc.rs [hunk 6]
```diff
@@ -48,6 +49,32 @@ impl JsonRpcClientWrapper {
...
+                        "ValidatorInfo",
+                        format!("{} is not found", account),
+                    )),
+                None => Err(Error::JsonRpcReadError(
+                    "ValidatorSet is not set",
+                    String::new(),
+                )),
+            },
+            Err(e) => Err(Error::JsonRpcReadError("ValidatorSet", e.to_string())),
+        }
```
Comment (gregnazario) [inline config/management/src/json_rpc.rs:75]: Nit: I mentioned this on Josh's too, I think you guys implemented the same function.  You can do a multi level match to make it a little cleaner

```
match validator_set {
  Ok(Some(validator_set)) => { ... }
  Ok(None) => { ... }
  Err(e) => { ... }
}
```

File: testsuite/tests/libratest/smoke_test.rs [hunk 7]
```diff
@@ -1035,14 +1042,18 @@ fn test_e2e_reconfiguration() {
...
-        .unwrap()
-        .to_string();
+    let peer_id = env.get_validator(0).unwrap().validator_peer_id().unwrap();
+    let op_tool = OperationalTool::new(
+        format!(
+            "http://127.0.0.1:{}",
+            env.validator_swarm.get_client_port(1)
+        ),
+        ChainId::test(),
+    );
```
Comment (gregnazario) [inline testsuite/tests/libratest/smoke_test.rs:1052]: Nit: Can probably make this a function?  I'm assuming we'll use it in more tests.

Also TIL that you can do `env.validator_swarm.get_client_port(#)`. I had been reading the configs to get that info in my test.

File: config/management/operational/src/command.rs (hunk 3 above)
Comment (zekun000) [reply config/management/operational/src/command.rs:130]: yeah I hate this

Comment (davidiw) [reply config/management/operational/src/command.rs:130]: A macro, where you pass in the right side of Command/CommandName and the return type. We'd need to convert the command to underscore case and then you could have this in 1 line.

**Pull request closed without merging**